src/ostree/ot-dump.c

```diff
@@ -407,6 +407,66 @@ ot_dump_summary_bytes (GBytes          *summary_bytes,
     }
 }
 
+static gint
+strptr_cmp (gconstpointer a,
+            gconstpointer b)
+{
+  const char *a_str = *((const char **) a);
+  const char *b_str = *((const char **) b);
+
+  return g_strcmp0 (a_str, b_str);
+}
+
+void
+ot_dump_summary_metadata_keys (GBytes *summary_bytes)
+{
+  g_autoptr(GVariant) summary = NULL;
+  g_autoptr(GVariant) metadata = NULL;
+
+  summary = g_variant_new_from_bytes (OSTREE_SUMMARY_GVARIANT_FORMAT,
+                                      summary_bytes, FALSE);
+  metadata = g_variant_get_child_value (summary, 1);
+
+  GVariantIter iter;
+  const char *key = NULL;
+  g_autoptr(GPtrArray) keys = g_ptr_array_new ();
+  g_variant_iter_init (&iter, metadata);
+  while (g_variant_iter_loop (&iter, "{&s@v}", &key, NULL))
+    g_ptr_array_add (keys, (gpointer) key);
+
+  g_ptr_array_sort (keys, strptr_cmp);
+  for (guint i = 0; i < keys-> len; i++)
+    {
+      key = keys->pdata[i];
+      g_print ("%s\n", key);
+    }
+}
+
+gboolean
+ot_dump_summary_metadata_key (GBytes      *summary_bytes,
+                              const char  *key,
+                              GError     **error)
+{
+  g_autoptr(GVariant) summary = NULL;
+  g_autoptr(GVariant) metadata = NULL;
+  g_autoptr(GVariant) value = NULL;
+
+  summary = g_variant_new_from_bytes (OSTREE_SUMMARY_GVARIANT_FORMAT,
```

Comment:
This is all fine as is but personally I've been trying to convert the whole codebase to declare-and-initialize because it's simply shorter, also matches how e.g. Rust works, and works better with `goto out` and not having implicitly nullable pointers.

I imagine you work on other codebases that still use C89 separation between declaration and init, but hopefully we can agree to keep this one using declare-and-initialize.

Reply:
That's fair. I split it for 2 reasons.

One is that I personally think sometimes it gets pretty busy with the initializer on the same line with an autoptr declaration. The other is that I have had gcc whine about using using uninitialized variables before when autoptr variables aren't explicitly initialized to `NULL`. Which seems like a compiler bug but there is some logic to it in certain scenarios.

Anyways, I'll be sure to do the full initialization next time.

Reply:
> The other is that I have had gcc whine about using using uninitialized variables before when autoptr variables aren't explicitly initialized to NULL. 

I'm talking about declare-and-initialize, so there's no point at which it's set to `NULL`.

Reply:
I hope I'm remembering this right, but I am talking about declare-and-initialize of a `g_autoptr`. I believe what happened is that the compiler decided that the initialization code could be run later, but then the `__attribute__((cleanup(func)))` code would warn that it was potentially using an uninitialized variable. As as said, likely a gcc bug since the presence of the cleanup attribute means it shouldn't split the initializer. I've only seen it a few times, though.